### PR TITLE
Instrumentation on HTTP exchanges

### PR DESCRIPTION
**Pull request by nyarly**
Sous as a client of Singularity and Docker registries,
as well as Sous as a server to itself (and any other API clients.)

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 10024 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: util/logging/messages/clienthttpresponse.go [hunk 1]
```diff
@@ -10,78 +10,132 @@ import (
...
+
+		// put the metrics into 3 buckets:
+		// by "resource family" - a way of aggregating URLs
+		// by hostname (with s/./_/g for graphite's comfort)
+		// by the pair of (resource, host)
+		return []string{
+			fmt.Sprintf("%s.%s", base, msg.resourceFamily),
+			fmt.Sprintf("%s.%s", base, host),
+			fmt.Sprintf("%s.%s.%s", base, host, msg.resourceFamily),
+		}
```
Comment (samsalisbury) [inline util/logging/messages/clienthttpresponse.go]: Sounds useful

File: util/logging/messages/clienthttpresponse.go [hunk 2]
```diff
@@ -10,78 +10,132 @@ import (
...
+		metrics.UpdateSample(name, msg.requestSize)
+	}
 }
 
-func (msg *clientHTTPResponse) EachField(f logging.FieldReportFn) {
-	f("@loglov3-otl", "http-v1")
-	f("incoming", false)
+func (msg *httpLogEntry) EachField(f logging.FieldReportFn) {
+	f("@loglov3-otl", "sous-http-v1")
+	f("resource-family", msg.resourceFamily)
```
Comment (samsalisbury) [inline util/logging/messages/clienthttpresponse.go]: Good name

File: util/restful/router.go [hunk 3]
```diff
@@ -54,9 +58,49 @@ func (xlog *ExchangeLogger) Exchange() (data interface{}, status int) {
...
+	lrw.statusCode = status
+	lrw.ResponseWriter.WriteHeader(status)
+}
+
+// Write logs the response. If ContentLength is set, we use that, otherwise we report a 0 length.
+// Unfortunately, ResponseWriter's contract makes it impossible to get the true content length.
+func (lrw loggingResponseWriter) Write(b []byte) (int, error) {
+	n, err := lrw.ResponseWriter.Write(b)
+
+	// ParseInt returns 0 and an syntax error if the provided string doesn't parse well.
```
Comment (samsalisbury) [inline util/restful/router.go]: Appreciate this comment for swallowing the error.

**Pull request merged**